submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,312 @@
+    Spec               ClusterEgressIPSpec `json:"spec"`
+
+    // Observed status of NamespaceGlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type ClusterEgressIPSpec struct {
+    // The number of globalIP's requested. Globalnet Controller will allocate the requested
+    // number of contiguous GlobalIPs for this ClusterGlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1.
+    // +optional
+    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from all Namespaces.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+type NamespaceGlobalEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of NamespaceGlobalEgressIP object.
+    Spec               NamespaceEgressIPSpec `json:"spec"`
+
+    // Observed status of NamespaceGlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type NamespaceEgressIPSpec struct {
+    // The number of globalIP's requested. Globalnet Controller will allocate the requested
+    // number of contiguous GlobalIPs for this NamespaceGlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1.
+    // +optional
+    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the namespace in which the
+    // CR is created.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+const (
+    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
+    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"
+
+    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
+    GlobalEgressIPError GlobalEgressIPStatus = "Error"
+)
+
+type EgressIPStatus struct {
+    // Status is one of {"Success", "Error"}
+    Status GlobalEgressIPStatus `json:"status,omitempty"`
+
+    // +optional
+    Message *string `json:"message,omitempty"`
+
+    // When more than a single GlobalIP is allocated, the starting GlobalIP allocated to this object.
+    StartIP string `json:"globalIPs"`
+
+    // When more than a single GlobalIP is allocated, the last GlobalIP (inclusive) allocated to this object.
+    EndIP string `json:"globalIPs"`
+
+    // The Namespaces to which the GlobalIPs are applied.
+    Namespaces []string `json:"namespaces"`
+
+    // The Pods to which the GlobalIPs are applied.
+    Pods []string `json:"pods"`
```

Comment:
I think we don't need to provide this information, it could be helpful for debugging but it would create a lot of work on the API side, and kind of overrides my comments in  L170, I hadn't noticed this.

We can add debug traces to this, ... but think of services, they provide no list of backend pods , just the list of IPs via a separate object (Endpoint).

I propose we remove Namespaces & Pods to avoid too much API trashing which would be a scalability issue (lots of pods come and go)

Reply:
I tend to agree about the pods part, namespaces probably dont churn as much.

IMHO it's better to annotate the pods (and maybe namespaces?) with the global policies in effect.

For first stage I'd remove these fields altogether though as I believe there's enough work as is without these aspects.
As Miguel proposed, this can be logged on the globalnet pod for now.

Reply:
Done, removed both of them. Infact I also added a note in L184 that we will no longer be using annotations on the Pods/Services as we are seeing issues with some services managed by certain controllers which periodically remove the annotation.